python/ray/experimental/async_plasma.py

```diff
@@ -183,30 +17,45 @@ def __init__(self, loop, worker):
         super().__init__()
         self._loop = loop
         self._worker = worker
-        self._waiting_dict = {}
+        self._waiting_dict = defaultdict(list)
 
     def process_notifications(self, messages):
         """Process notifications."""
         for object_id, object_size, metadata_size in messages:
             if object_size > 0 and object_id in self._waiting_dict:
-                linked_list = self._waiting_dict[object_id]
-                self._complete_future(linked_list)
+                self._complete_future(object_id)
 
     def close(self):
         """Clean up this handler."""
-        for linked_list in self._waiting_dict.values():
-            linked_list.cancel()
-        # All cancelled linked lists should have callbacks to removed itself
-        # from the waiting dict. However, these callbacks are scheduled in
-        # an event loop, so we don't check them now.
-
-    def _unregister_callback(self, fut):
-        del self._waiting_dict[fut.object_id]
-
-    def _complete_future(self, fut):
-        obj = self._worker.get_objects([ray.ObjectID(
-            fut.object_id.binary())])[0]
-        fut.set_result(obj)
+        for futures in self._waiting_dict.values():
+            for fut in futures:
+                fut.cancel()
+
+    def _complete_future(self, ray_object_id):
+        # TODO(ilr): Consider race condition between this method and as_future
```

Comment:
```suggestion
        # TODO(ilr): Consider race condition between this method and as_future.
```
Can you also please describe the race condition here for the benefit of others?

Reply:
Done!